### PR TITLE
Improve the extension's usage of Sentry

### PR DESCRIPTION
**Pull request by liamoneill**
This PR aims to improve our usage of Sentry inside the extension. Sentry was originally integrated after the fact into the existing logging framework and we weren't taking advantage of Sentry's event structure and instead logging free text stack breadcrumbs and backtraces.

This PR has the following changes:
- Sends through MonoLog's context values in Sentry Error's extras and Breadcrumb's metadata.
- Sends through `\Throwable` instances in the `exception` field inside MonoLog's context as a Sentry exception. This allows Sentry to display a structured stacktrace in it's UI rather than a freetext backtrace as we currently have.
- Sends through the extension's version in Sentry's `release` tag.
- Adapts some of the existing logger usage to take advantage of these changes.

#### Before
```php
$this->logger->debug('Unexpected error while processing event');
$this->logger->error($exception);
```
#### After
```php
// Note that `exception` is a special field that if present will be hoisted into the Sentry Event's `exception` field.
$this->logger->error('Unexpected error while processing event', ['exception' => $exception]);
```

### CODE REVIEW

File: Model/Logger/SentryHandler.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+use Monolog\Handler\AbstractProcessingHandler;
+use Monolog\Logger;
+use Sentry\Event;
+use Sentry\EventHint;
+use Sentry\Severity;
+use Sentry\State\HubInterface;
+use Sentry\State\Scope;
+
+// Adapted from https://github.com/getsentry/sentry-php/blob/master/src/Monolog/Handler.php
```
Comment (dylan-chong) [inline Model/Logger/SentryHandler.php:15]: Why do we need to do this?

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,4 @@
 .vagrant
 /vagrant/.env
 /vagrant/.composer
+.vscode/
```
Comment (dylan-chong) [inline .gitignore:6]: 😄

**Comment by dylan-chong** [comment]
Got an error running vagrant provision today

```
    default: Performing first time setup of Magento
    default: + docker_compose exec php-fpm sh -c '
    default:             sudo -u www-data php bin/magento setup:install \
    default:                 --base-url="http://${MAGENTO_WEB_ADDRESS}:${MAGENTO_WEB_PORT}/" \
    default:                 --db-host="${MYSQL_HOST}" \
    default:                 --db-name="${MYSQL_DATABASE}" \
    default:                 --db-user="${MYSQL_USER}" \
    default:                 --db-password="${MYSQL_PASSWORD}" \
    default:                 --admin-firstname="${MAGENTO_ADMIN_FIRSTNAME}" \
    default:                 --admin-lastname="${MAGENTO_ADMIN_LASTNAME}" \
    default:                 --admin-email="${MAGENTO_ADMIN_EMAIL}" \
    default:                 --admin-user="${MAGENTO_ADMIN_USER}" \
    default:                 --admin-password="${MAGENTO_ADMIN_PASSWORD}" \
    default:                 --language="${MAGENTO_LANGUAGE}" \
    default:                 --currency="${MAGENTO_CURRENCY}" \
    default:                 --timezone="${MAGENTO_TIMEZONE}"
    default:             '
    default: + cd -- .
    default: + [[ exec == \e\x\e\c ]]
    default: + [[ ! -t 1 ]]
    default: + shift
    default: + docker-compose exec -T php-fpm sh -c '
    default:             sudo -u www-data php bin/magento setup:install \
    default:                 --base-url="http://${MAGENTO_WEB_ADDRESS}:${MAGENTO_WEB_PORT}/" \
    default:                 --db-host="${MYSQL_HOST}" \
    default:                 --db-name="${MYSQL_DATABASE}" \
    default:                 --db-user="${MYSQL_USER}" \
    default:                 --db-password="${MYSQL_PASSWORD}" \
    default:                 --admin-firstname="${MAGENTO_ADMIN_FIRSTNAME}" \
    default:                 --admin-lastname="${MAGENTO_ADMIN_LASTNAME}" \
    default:                 --admin-email="${MAGENTO_ADMIN_EMAIL}" \
    default:                 --admin-user="${MAGENTO_ADMIN_USER}" \
    default:                 --admin-password="${MAGENTO_ADMIN_PASSWORD}" \
    default:                 --language="${MAGENTO_LANGUAGE}" \
    default:                 --currency="${MAGENTO_CURRENCY}" \
    default:                 --timezone="${MAGENTO_TIMEZONE}"
    default:             '
    default: Starting Magento installation:
    default: File permissions check...
    default: [Progress: 1 / 1001]
    default: Required extensions check...
    default: [Progress: 2 / 1001]
    default: Enabling Maintenance Mode...
    default: [Progress: 3 / 1001]
    default: Installing deployment configuration...
    default: [Progress: 4 / 1001]
    default: Installing database schema:
    default: In AbstractFactory.php line 131:
    default:
    default:   Type Error occurred when creating object: SolveData\Events\Model\Logger\Sen
    default:   tryHubManager
    default:
    default:
    default: setup:install [--amqp-host AMQP-HOST] [--amqp-port AMQP-PORT] [--amqp-user AMQP-USER] [--amqp-password AMQP-PASSWORD] [--amqp-virtualhost AMQP-VIRTUALHOST] [--amqp-ssl AMQP-SSL] [--amqp-ssl-options AMQP-SSL-OPTIONS] [--enable-debug-logging ENABLE-DEBUG-LOGGING] [--enable-syslog-logging ENABLE-SYSLOG-LOGGING] [--backend-frontname BACKEND-FRONTNAME] [--consumers-wait-for-messages CONSUMERS-WAIT-FOR-MESSAGES] [--key KEY] [--db-host DB-HOST] [--db-name DB-NAME] [--db-user DB-USER] [--db-engine DB-ENGINE] [--db-password DB-PASSWORD] [--db-prefix DB-PREFIX] [--db-model DB-MODEL] [--db-init-statements DB-INIT-STATEMENTS] [-s|--skip-db-validation] [--http-cache-hosts HTTP-CACHE-HOSTS] [--db-ssl-key DB-SSL-KEY] [--db-ssl-cert DB-SSL-CERT] [--db-ssl-ca DB-SSL-CA] [--db-ssl-verify] [--session-save SESSION-SAVE] [--session-save-redis-host SESSION-SAVE-REDIS-HOST] [--session-save-redis-port SESSION-SAVE-REDIS-PORT] [--session-save-redis-password SESSION-SAVE-REDIS-PASSWORD] [--session-save-redis-timeout SESSION-SAVE-REDIS-TIMEOUT] [--session-save-redis-persistent-id SESSION-SAVE-REDIS-PERSISTENT-ID] [--session-save-redis-db SESSION-SAVE-REDIS-DB] [--session-save-redis-compression-threshold SESSION-SAVE-REDIS-COMPRESSION-THRESHOLD] [--session-save-redis-compression-lib SESSION-SAVE-REDIS-COMPRESSION-LIB] [--session-save-redis-log-level SESSION-SAVE-REDIS-LOG-LEVEL] [--session-save-redis-max-concurrency SESSION-SAVE-REDIS-MAX-CONCURRENCY] [--session-save-redis-break-after-frontend SESSION-SAVE-REDIS-BREAK-AFTER-FRONTEND] [--session-save-redis-break-after-adminhtml SESSION-SAVE-REDIS-BREAK-AFTER-ADMINHTML] [--session-save-redis-first-lifetime SESSION-SAVE-REDIS-FIRST-LIFETIME] [--session-save-redis-bot-first-lifetime SESSION-SAVE-REDIS-BOT-FIRST-LIFETIME] [--session-save-redis-bot-lifetime SESSION-SAVE-REDIS-BOT-LIFETIME] [--session-save-redis-disable-locking SESSION-SAVE-REDIS-DISABLE-LOCKING] [--session-save-redis-min-lifetime SESSION-SAVE-REDIS-MIN-LIFETIME] [--session-save-redis-max-lifetime SESSION-SAVE-REDIS-MAX-LIFETIME] [--session-save-redis-sentinel-master SESSION-SAVE-REDIS-SENTINEL-MASTER] [--session-save-redis-sentinel-servers SESSION-SAVE-REDIS-SENTINEL-SERVERS] [--session-save-redis-sentinel-verify-master SESSION-SAVE-REDIS-SENTINEL-VERIFY-MASTER] [--session-save-redis-sentinel-connect-retires SESSION-SAVE-REDIS-SENTINEL-CONNECT-RETIRES] [--cache-backend CACHE-BACKEND] [--cache-backend-redis-server CACHE-BACKEND-REDIS-SERVER] [--cache-backend-redis-db CACHE-BACKEND-REDIS-DB] [--cache-backend-redis-port CACHE-BACKEND-REDIS-PORT] [--cache-backend-redis-password CACHE-BACKEND-REDIS-PASSWORD] [--cache-backend-redis-compress-data CACHE-BACKEND-REDIS-COMPRESS-DATA] [--cache-backend-redis-compression-lib CACHE-BACKEND-REDIS-COMPRESSION-LIB] [--cache-id-prefix CACHE-ID-PREFIX] [--page-cache PAGE-CACHE] [--page-cache-redis-server PAGE-CACHE-REDIS-SERVER] [--page-cache-redis-db PAGE-CACHE-REDIS-DB] [--page-cache-redis-port PAGE-CACHE-REDIS-PORT] [--page-cache-redis-password PAGE-CACHE-REDIS-PASSWORD] [--page-cache-redis-compress-data PAGE-CACHE-REDIS-COMPRESS-DATA] [--page-cache-redis-compression-lib PAGE-CACHE-REDIS-COMPRESSION-LIB] [--page-cache-id-prefix PAGE-CACHE-ID-PREFIX] [--lock-provider LOCK-PROVIDER] [--lock-db-prefix LOCK-DB-PREFIX] [--lock-zookeeper-host LOCK-ZOOKEEPER-HOST] [--lock-zookeeper-path LOCK-ZOOKEEPER-PATH] [--lock-file-path LOCK-FILE-PATH] [--base-url BASE-URL] [--language LANGUAGE] [--timezone TIMEZONE] [--currency CURRENCY] [--use-rewrites USE-REWRITES] [--use-secure USE-SECURE] [--base-url-secure BASE-URL-SECURE] [--use-secure-admin USE-SECURE-ADMIN] [--admin-use-security-key ADMIN-USE-SECURITY-KEY] [--admin-user [ADMIN-USER]] [--admin-password [ADMIN-PASSWORD]] [--admin-email [ADMIN-EMAIL]] [--admin-firstname [ADMIN-FIRSTNAME]] [--admin-lastname [ADMIN-LASTNAME]] [--cl
    default: eanup-database] [--sales-order-increment-prefix SALES-ORDER-INCREMENT-PREFIX] [--use-sample-data] [--enable-modules [ENABLE-MODULES]] [--disable-modules [DISABLE-MODULES]] [--convert-old-scripts [CONVERT-OLD-SCRIPTS]] [-i|--interactive] [--safe-mode [SAFE-MODE]] [--data-restore [DATA-RESTORE]] [--dry-run [DRY-RUN]] [--magento-init-params MAGENTO-INIT-PARAMS]
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

```

hmm

**Review by dylan-chong — APPROVED** [review]
Looks great to me

**Comment by liamoneill** [comment]
> Got an error running vagrant provision today

Turns out this error was due to vagrant provision not recompiling the extension. This is a minor issue with our localdev tooling, the work around is to run `./tools.sh recompile` when that error is encountered with `vagrant provision`.

**Pull request merged**